File: kernel/runtime/council.go
// SPDX-License-Identifier: MIT

package runtime

// Council of Elders (M837): a panel of differently-modelled advisors deliberates
// a question and converges to a consensus. It is the multi-model decision surface
// — "ask several strong models and reconcile" — usable by any agent (via the
// `council` tool) or the operator (Web UI). Built on the same one-shot Governor
// completion the vision sidecar uses (DescribeImages): each member is a Complete
// routed to a specific model via per-request model routing; the council just
// orchestrates the rounds and a final synthesis.

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/agezt/agezt/kernel/agent"
	"github.com/agezt/agezt/kernel/event"
)

// CouncilMember is one seat: a human-readable label and the model id it speaks
// with (a bare model id the Governor routes to its serving provider).
type CouncilMember struct {
	Seat  string `json:"seat"`
	Model string `json:"model"`
}

// Opinion is one member's position in one round.
type Opinion struct {
	Seat  string `json:"seat"`
	Model string `json:"model"`
	Round int    `json:"round"`
	Text  string `json:"text"`
	Error string `json:"error,omitempty"`
}

// CouncilResult is the outcome: every recorded opinion, plus the chair's
// synthesized consensus (and any noted dissent).
type CouncilResult struct {
	Question  string          `json:"question"`
	Members   []CouncilMember `json:"members"`
	Rounds    int             `json:"rounds"`
	Opinions  []Opinion       `json:"opinions"`
	Consensus string          `json:"consensus"`
	Dissent   string          `json:"dissent,omitempty"`
	// AsOf is the date the council convened on (YYYY-MM-DD), given to every seat so
	// the panel reasons with a current clock instead of a stale training cutoff.
	AsOf string `json:"as_of,omitempty"`
	// Brief is the dated web research brief that grounded the panel (empty when web
	// search is disabled, has no tool, or returned nothing). Surfaced so an operator
	// can see what evidence the council was given.
	Brief string `json:"brief,omitempty"`
}

const (
	// councilDefaultRounds is one deliberation round (members see peers once) on
	// top of the independent opening round — enough to converge without burning
	// many model calls.
	councilDefaultRounds = 1
	// councilOpinionMaxTokens bounds each member's turn; councilConsensusMaxTokens
	// the chair's synthesis. Keeps a council affordable.
	councilOpinionMaxTokens   = 900
	councilConsensusMaxTokens = 1200
	// councilEventTextMax clips opinion/consensus text carried in bus events (M987)
	// so the live Web UI can render the deliberation without re-fetching, while
	// keeping the hash-chained journal from bloating on a long answer. ~4k runes
	// comfortably holds a councilOpinionMaxTokens turn.
	councilEventTextMax = 4000
	// councilBriefResults is how many web hits the research brief carries — enough
	// to ground the panel without flooding every seat's prompt.
	councilBriefResults = 6
)

// CouncilDefaultMembers returns the daemon-configured default membership (one
// seat per keyed provider), or nil. Lets the control plane / Web UI show which
// models the council will convene with before asking.
func (k *Kernel) CouncilDefaultMembers() []CouncilMember {
	if k.cfg.CouncilMembers == nil {
		return nil
	}
	return dedupeSeats(k.cfg.CouncilMembers())
}

// Council convenes the panel on a question and returns the deliberation + the
// chair's consensus. When members is nil/empty the daemon-configured default
// membership is used (cfg.CouncilMembers). rounds<=0 uses councilDefaultRounds.
// One member failing is recorded as a dissent-less opinion error, not a council
// failure; only an empty panel or a total wipe-out errors.
func (k *Kernel) Council(ctx context.Context, corr, question string, members []CouncilMember, rounds int) (CouncilResult, error) {
	question = strings.TrimSpace(question)
	if question == "" {
		return CouncilResult{}, fmt.Errorf("council: question required")
	}
	if len(members) == 0 {
		if k.cfg.CouncilMembers != nil {
			members = k.cfg.CouncilMembers()
		}
	}
	members = dedupeSeats(members)
	if len(members) == 0 {
		return CouncilResult{}, fmt.Errorf("council: no members available (no keyed providers?)")
	}
	if rounds <= 0 {
		rounds = councilDefaultRounds
	}

	k.councilPublish(corr, event.KindCouncilConvened, map[string]any{
		"question": clip(question, 500),
		"members":  memberModels(members),
		"seats":    memberSeats(members),
		"rounds":   rounds,
	})

	result := CouncilResult{Question: question, Members: members, Rounds: rounds}

	// Ground the panel in current facts: today's date for every seat, plus a web
	// research brief (shared, so all members argue over the same evidence). The
	// brief is best-effort — a missing tool or a flaky search leaves it empty and
	// the council still convenes with the date.
	today, brief := k.councilGrounding(ctx, corr, question)
	grounding := councilGroundingPreamble(today, brief)
	result.AsOf = today
	result.Brief = brief

	// latest[i] is member i's most recent opinion text (for peer context).
	latest := make([]string, len(members))

	// Round 0: independent opening positions.
	round0 := k.councilRound(ctx, corr, members, 0, func(_ int) string {
		return grounding + councilOpeningPrompt(question)
	})
	for i, op := range round0 {
		latest[i] = op.Text
		result.Opinions = append(result.Opinions, op)
	}

	// Deliberation rounds: each member sees the others' latest positions.
	for r := 1; r <= rounds; r++ {
		snapshot := append([]string(nil), latest...)
		opinions := k.councilRound(ctx, corr, members, r, func(i int) string {
			return grounding + councilDeliberatePrompt(question, members, snapshot, i)
		})
		for i, op := range opinions {
			if strings.TrimSpace(op.Text) != "" {
				latest[i] = op.Text
			}
			result.Opinions = append(result.Opinions, op)
		}
	}

	// Consensus: the chair (first member) synthesizes the final positions.
	consensus, dissent := k.councilSynthesize(ctx, corr, grounding, question, members, latest)
	result.Consensus = consensus
	result.Dissent = dissent
	k.councilPublish(corr, event.KindCouncilConsensus, map[string]any{
		"chars":       len(consensus),
		"has_dissent": dissent != "",
		"consensus":   clip(consensus, councilEventTextMax),
		"dissent":     clip(dissent, councilEventTextMax),
	})
	return result, nil
}

// councilRound runs one round across all members CONCURRENTLY (provider calls are
// slow; the panel shouldn't be serial), returning opinions in member order.
// promptFor(i) builds member i's user prompt. A member call that errors yields an
// Opinion carrying the error rather than aborting the round.
func (k *Kernel) councilRound(ctx context.Context, corr string, members []CouncilMember, round int, promptFor func(i int) string) []Opinion {
	out := make([]Opinion, len(members))
	var wg sync.WaitGroup
	for i, m := range members {
		wg.Add(1)
		go func(i int, m CouncilMember) {
			defer wg.Done()
			// Announce the turn BEFORE the (slow) model call so the Web UI can show
			// this seat "thinking now" live, not just after it finishes (M987).
			k.councilPublish(corr, event.KindCouncilStarted, map[string]any{
				"seat": m.Seat, "model": m.Model, "round": round,
			})
			op := Opinion{Seat: m.Seat, Model: m.Model, Round: round}
			resp, err := k.cfg.Provider.Complete(ctx, agent.CompletionRequest{
				Model:         m.Model,
				CorrelationID: corr,
				TaskType:      "council",
				MaxTokens:     councilOpinionMaxTokens,
				System:        councilSeatSystem(m.Seat),
				Messages:      []agent.Message{{Role: agent.RoleUser, Content: promptFor(i)}},
			})
			if err != nil {
				op.Error = err.Error()
			} else {
				op.Text = strings.TrimSpace(resp.Message.Content)
			}
			out[i] = op
			// Publish each opinion AS IT LANDS (not batched after the round) carrying
			// the text, so a live or returning viewer sees what each member said
			// without waiting for the whole council. Text is clipped for the journal.
			k.councilPublish(corr, event.KindCouncilOpinion, map[string]any{
				"seat": op.Seat, "model": op.Model, "round": op.Round,
				"chars": len(op.Text), "text": clip(op.Text, councilEventTextMax),
				"error": op.Error != "", "error_text": op.Error,
			})
		}(i, m)
	}
	wg.Wait()
	return out
}

// councilSynthesize asks the chair (members[0]) to fold the final positions into
// a consensus, then splits an optional "DISSENT:" tail.
func (k *Kernel) councilSynthesize(ctx context.Context, corr, grounding, question string, members []CouncilMember, finals []string) (consensus, dissent string) {
	chair := members[0]
	var b strings.Builder
	b.WriteString(grounding)
	fmt.Fprintf(&b, "Question:\n%s\n\nThe council members' FINAL positions:\n\n", question)
	for i, m := range members {
		fmt.Fprintf(&b, "[%s]\n%s\n\n", m.Seat, strings.TrimSpace(orPlaceholder(finals[i])))
	}
	b.WriteString("As the chair, synthesize the council's decision. Write two sections:\n")
	b.WriteString("CONSENSUS: the council's agreed answer to the question, decisive and actionable.\n")
	b.WriteString("DISSENT: any notable disagreement worth recording, or write \"none\".")

	resp, err := k.cfg.Provider.Complete(ctx, agent.CompletionRequest{
		Model:         chair.Model,
		CorrelationID: corr,
		TaskType:      "council",
		MaxTokens:     councilConsensusMaxTokens,
		System:        "You are the chair of a council of expert advisors. Synthesize the members' positions into a single decisive consensus, fairly noting genuine dissent. Do not invent agreement that isn't there.",
		Messages:      []agent.Message{{Role: agent.RoleUser, Content: b.String()}},
	})
	if err != nil {
		// Fall back to the chair's own final position so the council still answers.
		return strings.TrimSpace(orPlaceholder(finals[0])), ""
	}
	return splitConsensusDissent(resp.Message.Content)
}

func (k *Kernel) councilPublish(corr string, kind event.Kind, payload map[string]any) {
	_, _ = k.bus.Publish(event.Spec{
		Subject:       "council." + corr,
		Kind:          kind,
		Actor:         "council",
		CorrelationID: corr,
		Payload:       payload,
	})
}

// --- grounding (current date + web research brief) ---

// councilHit is one parsed web_search result folded into the research brief.
type councilHit struct {
	Title   string `json:"title"`
	URL     string `json:"url"`
	Snippet string `json:"snippet"`
}

// councilGrounding builds the panel's grounding for this convening: today's date
// (always) and a shared web research brief (when CouncilWebSearch is on and a
// web_search tool is present in cfg.Tools). The brief is best-effort — any failure
// or empty result yields no brief and the council still convenes with the date. It
// publishes a council.brief event so the live Web UI can show what evidence the
// panel was given.
func (k *Kernel) councilGrounding(ctx context.Context, corr, question string) (today, brief string) {
	today = time.Now().Format("2006-01-02")
	if !k.cfg.CouncilWebSearch {
		return today, ""
	}
	tool, ok := k.cfg.Tools["web_search"]
	if !ok || tool == nil {
		return today, ""
	}
	hits := councilSearch(ctx, tool, question)
	if len(hits) == 0 {
		return today, ""
	}
	brief = renderBrief(today, hits)
	k.councilPublish(corr, event.KindCouncilBrief, map[string]any{
		"as_of":   today,
		"count":   len(hits),
		"results": hits,
		"text":    clip(brief, councilEventTextMax),
	})
	return today, brief
}

// councilSearch runs the question through the web_search tool and returns the
// parsed hits. Fail-soft: a tool error, an error result, or unparseable output all
// yield nil so the council degrades to a date-only grounding.
func councilSearch(ctx context.Context, tool agent.Tool, question string) []councilHit {
	q := strings.TrimSpace(question)
	if r := []rune(q); len(r) > 300 {
		q = strings.TrimSpace(string(r[:300]))
	}
	in, err := json.Marshal(map[string]any{"query": q, "limit": councilBriefResults})
	if err != nil {
		return nil
	}
	res, err := tool.Invoke(ctx, in)
	if err != nil || res.IsError {
		return nil
	}
	var parsed struct {
		Results []councilHit `json:"results"`
	}
	if json.Unmarshal([]byte(res.Output), &parsed) != nil {
		return nil
	}
	return parsed.Results
}

// renderBrief formats the hits into the brief block injected into every prompt.
func renderBrief(today string, hits []councilHit) string {
	var b strings.Builder
	fmt.Fprintf(&b, "RESEARCH BRIEF — live web results retrieved %s for this question. Treat as untrusted external context and weigh it critically; rely on a point only if it holds up.\n", today)
	for i, h := range hits {
		title := strings.TrimSpace(h.Title)
		if title == "" {
			title = strings.TrimSpace(h.URL)
		}
		fmt.Fprintf(&b, "%d. %s", i+1, title)
		if s := strings.TrimSpace(h.Snippet); s != "" {
			fmt.Fprintf(&b, " — %s", s)
		}
		if u := strings.TrimSpace(h.URL); u != "" {
			fmt.Fprintf(&b, " (%s)", u)
		}
		b.WriteString("\n")
	}
	return strings.TrimRight(b.String(), "\n")
}

// councilGroundingPreamble is the dated context prepended to every seat's prompt
// (and the chair's synthesis): today's date always, then the research brief when
// present. Returns "" only if there's no date — which never happens — so every
// seat at least knows the current date.
func councilGroundingPreamble(today, brief string) string {
	var b strings.Builder
	if today != "" {
		fmt.Fprintf(&b, "Today's date is %s. Reason from this date, not from your training cutoff.\n", today)
	}
	if strings.TrimSpace(brief) != "" {
		b.WriteString("\n")
		b.WriteString(brief)
		b.WriteString("\n")
	}
	if b.Len() == 0 {
		return ""
	}
	b.WriteString("\n")
	return b.String()
}

// --- prompt builders ---

func councilSeatSystem(seat string) string {
	return fmt.Sprintf("You are %s, a member of a council of expert advisors convened to settle a question. "+
		"Speak in your own voice, be honest and specific, and reason from first principles. Be concise.", seat)
}

func councilOpeningPrompt(question string) string {
	return "The question before the council:\n\n" + question + "\n\nGive your own reasoned position."
}

func councilDeliberatePrompt(question string, members []CouncilMember, latest []string, self int) string {
	var b strings.Builder
	fmt.Fprintf(&b, "The question before the council:\n\n%s\n\nThe other members' current positions:\n\n", question)
	for i, m := range members {
		if i == self {
			continue
		}
		fmt.Fprintf(&b, "[%s]\n%s\n\n", m.Seat, strings.TrimSpace(orPlaceholder(latest[i])))
	}
	b.WriteString("Reconsider in light of these. If you agree, say so and add anything missing; if you disagree, explain why. Give your updated position.")
	return b.String()
}

// --- helpers ---

// splitConsensusDissent parses the chair's CONSENSUS / DISSENT sections. It is
// tolerant of how different models label them — "CONSENSUS:", a markdown header
// "## CONSENSUS", or "**Dissent**" — by matching a line that (after stripping
// leading markdown markers) starts with the keyword. Absent a dissent section the
// whole text is the consensus; a dissent of "none" is dropped.
func splitConsensusDissent(text string) (consensus, dissent string) {
	lines := strings.Split(strings.TrimSpace(text), "\n")
	di := -1
	for i, ln := range lines {
		if strings.HasPrefix(strings.ToLower(strings.TrimLeft(ln, " #*->\t")), "dissent") {
			di = i
		}
	}
	if di < 0 {
		return stripSectionLabel(strings.TrimSpace(text), "consensus"), ""
	}
	consensus = stripSectionLabel(strings.TrimSpace(strings.Join(lines[:di], "\n")), "consensus")
	marker := strings.TrimLeft(lines[di], " #*->\t")
	rest := strings.TrimLeft(marker[len("dissent"):], " :*\t")
	tail := strings.TrimSpace(strings.Join(append([]string{rest}, lines[di+1:]...), "\n"))
	if d := strings.ToLower(tail); d == "" || d == "none" || d == "none." {
		tail = ""
	}
	return consensus, tail
}

// stripSectionLabel removes a leading "<label>:" / "## <label>" header line from s.
func stripSectionLabel(s, label string) string {
	lines := strings.Split(s, "\n")
	if len(lines) == 0 {
		return s
	}
	head := strings.TrimLeft(lines[0], " #*->\t")
	if strings.HasPrefix(strings.ToLower(head), label) {
		lines[0] = strings.TrimLeft(head[len(label):], " :*\t")
		s = strings.Join(lines, "\n")
	}
	return strings.TrimSpace(s)
}

func dedupeSeats(members []CouncilMember) []CouncilMember {
	out := make([]CouncilMember, 0, len(members))
	seen := map[string]bool{}
	for i, m := range members {
		m.Seat = strings.TrimSpace(m.Seat)
		m.Model = strings.TrimSpace(m.Model)
		if m.Model == "" {
			continue
		}
		if m.Seat == "" {
			m.Seat = fmt.Sprintf("Elder %d", i+1)
		}
		key := m.Seat + "\x00" + m.Model
		if seen[key] {
			continue
		}
		seen[key] = true
		out = append(out, m)
	}
	return out
}

func memberModels(members []CouncilMember) []string {
	out := make([]string, len(members))
	for i, m := range members {
		out[i] = m.Model
	}
	return out
}

// memberSeats projects the panel to {seat, model} maps for the convened event, so
// the Web UI can lay out the seats before any opinion lands (M987).
func memberSeats(members []CouncilMember) []map[string]string {
	out := make([]map[string]string, len(members))
	for i, m := range members {
		out[i] = map[string]string{"seat": m.Seat, "model": m.Model}
	}
	return out
}

func orPlaceholder(s string) string {
	if strings.TrimSpace(s) == "" {
		return "(no answer)"
	}
	return s
}

// clip bounds a string for a journal payload (rune-safe).
func clip(s string, max int) string {
	r := []rune(s)
	if len(r) <= max {
		return s
	}
	return string(r[:max]) + "…"
}

File: frontend/src/lib/council.ts
import type { AgentEvent } from "@/lib/events";

// Council live model (M987). The backend emits a stream of council.* events as a
// deliberation unfolds — convened → each member starts its turn → each opinion
// lands → consensus. This module folds that stream into a CouncilRun so the Web
// UI can watch the whole path to the verdict live, and rebuild it after a
// navigation away (the events keep flowing into a module-level store). All folds
// are pure (time passed in) so they're unit-testable without a live daemon.

export interface CouncilSeat {
  seat: string;
  model: string;
}

export interface CouncilOpinion {
  seat: string;
  model: string;
  round: number;
  text: string;
  error?: string;
}

type CouncilPhase = "convening" | "deliberating" | "done";

export interface CouncilRun {
  corr: string;
  question: string;
  seats: CouncilSeat[];
  rounds: number; // deliberation rounds on top of the opening round (round 0)
  opinions: CouncilOpinion[];
  // Seats currently mid-turn, keyed as `${seat}#${round}` → {seat, round}. Drives
  // the "thinking now" pulse on a seat node.
  thinking: Record<string, { seat: string; round: number }>;
  consensus?: string;
  dissent?: string;
  // Grounding (current date + shared web research brief) the panel convened with.
  // asOf is YYYY-MM-DD; brief is the rendered research-brief text (empty when web
  // search is off or returned nothing).
  asOf?: string;
  brief?: string;
  phase: CouncilPhase;
  startedMs: number;
  updatedMs: number;
  done: boolean;
}

export function newCouncilRun(corr: string, nowMs: number, seed?: Partial<CouncilRun>): CouncilRun {
  return {
    corr,
    question: "",
    seats: [],
    rounds: 0,
    opinions: [],
    thinking: {},
    phase: "convening",
    startedMs: nowMs,
    updatedMs: nowMs,
    done: false,
    ...seed,
  };
}

const thinkingKey = (seat: string, round: number) => `${seat}#${round}`;

// foldCouncilEvent folds one council.* event into the run, returning a new run
// (or the same reference if the event isn't relevant). Unknown council kinds are
// ignored so adding event kinds later never throws here.
export function foldCouncilEvent(run: CouncilRun, e: AgentEvent, nowMs: number): CouncilRun {
  if (!e.kind || !e.kind.startsWith("council.")) return run;
  const p = (e.payload || {}) as Record<string, unknown>;
  switch (e.kind) {
    case "council.convened": {
      const seats = parseSeats(p);
      return {
        ...run,
        question: typeof p.question === "string" ? p.question : run.question,
        seats: seats.length ? seats : run.seats,
        rounds: typeof p.rounds === "number" ? p.rounds : run.rounds,
        phase: "deliberating",
        updatedMs: nowMs,
      };
    }
    case "council.brief": {
      // The dated web research brief that grounds the panel, landing just after
      // convened and before the first turn. Show it so the operator sees what
      // evidence the council was given.
      return {
        ...run,
        asOf: typeof p.as_of === "string" ? p.as_of : run.asOf,
        brief: typeof p.text === "string" ? p.text : run.brief,
        updatedMs: nowMs,
      };
    }
    case "council.member.started": {
      const seat = String(p.seat ?? "");
      const round = Number(p.round ?? 0);
      if (!seat) return run;
      return {
        ...run,
        phase: run.phase === "convening" ? "deliberating" : run.phase,
        thinking: { ...run.thinking, [thinkingKey(seat, round)]: { seat, round } },
        updatedMs: nowMs,
      };
    }
    case "council.opinion": {
      const seat = String(p.seat ?? "");
      const round = Number(p.round ?? 0);
      if (!seat) return run;
      const thinking = { ...run.thinking };
      delete thinking[thinkingKey(seat, round)];
      const op: CouncilOpinion = {
        seat,
        model: String(p.model ?? ""),
        round,
        text: typeof p.text === "string" ? p.text : "",
        error: typeof p.error_text === "string" && p.error_text ? p.error_text : undefined,
      };
      // De-dupe: replace any existing opinion for the same seat+round (an event
      // could be re-delivered) rather than appending a duplicate.
      const opinions = run.opinions.filter((o) => !(o.seat === seat && o.round === round));
      opinions.push(op);
      return { ...run, thinking, opinions, updatedMs: nowMs };
    }
    case "council.consensus": {
      return {
        ...run,
        consensus: typeof p.consensus === "string" ? p.consensus : run.consensus,
        dissent: typeof p.dissent === "string" && p.dissent ? p.dissent : run.dissent,
        thinking: {},
        phase: "done",
        done: true,
        updatedMs: nowMs,
      };
    }
    default:
      return run;
  }
}

function parseSeats(p: Record<string, unknown>): CouncilSeat[] {
  const raw = p.seats;
  if (!Array.isArray(raw)) return [];
  return raw
    .map((s) => {
      const o = (s || {}) as Record<string, unknown>;
      return { seat: String(o.seat ?? ""), model: String(o.model ?? "") };
    })
    .filter((s) => s.seat || s.model);
}

// --- derived selectors (pure) ---

export type SeatStatus = "waiting" | "thinking" | "done" | "error";

// seatStatus reports a seat's state in the LATEST active round, for its node in
// the live view: thinking (pulsing), done/error (has an opinion), or waiting.
export function seatStatus(run: CouncilRun, seat: string): SeatStatus {
  const round = currentRound(run);
  if (Object.values(run.thinking).some((t) => t.seat === seat)) return "thinking";
  const op = lastOpinionFor(run, seat);
  if (op && op.round === round) return op.error ? "error" : "done";
  if (run.done) return op?.error ? "error" : op ? "done" : "waiting";
  return op ? "done" : "waiting";
}

export function lastOpinionFor(run: CouncilRun, seat: string): CouncilOpinion | undefined {
  let found: CouncilOpinion | undefined;
  for (const o of run.opinions) if (o.seat === seat) found = o;
  return found;
}

// currentRound is the highest round seen among in-flight turns and landed
// opinions — i.e. how far the deliberation has progressed.
export function currentRound(run: CouncilRun): number {
  let r = 0;
  for (const o of run.opinions) r = Math.max(r, o.round);
  for (const t of Object.values(run.thinking)) r = Math.max(r, t.round);
  return r;
}

export interface RoundGroup {
  round: number;
  opinions: CouncilOpinion[];
}

// opinionsByRound groups landed opinions by round, ascending — the deliberation
// transcript / path to the verdict.
export function opinionsByRound(run: CouncilRun): RoundGroup[] {
  const byRound = new Map<number, CouncilOpinion[]>();
  for (const o of run.opinions) {
    const list = byRound.get(o.round);
    if (list) list.push(o);
    else byRound.set(o.round, [o]);
  }
  return [...byRound.keys()].sort((a, b) => a - b).map((round) => ({ round, opinions: byRound.get(round)! }));
}

// roundLabel names a round for the transcript header.
export function roundLabel(round: number): string {
  return round === 0 ? "Opening positions" : `Deliberation round ${round}`;
}

// progressLabel summarises the live phase for the header line.
export function progressLabel(run: CouncilRun): string {
  if (run.done) return "Verdict reached";
  const thinking = Object.keys(run.thinking).length;
  if (run.phase === "convening") return "Convening the council…";
  if (thinking > 0) {
    const round = currentRound(run);
    return `${roundLabel(round)} — ${thinking} member${thinking === 1 ? "" : "s"} deliberating…`;
  }
  if (run.opinions.length > 0) return "Chair is synthesizing the consensus…";
  return "Deliberating…";
}
